src/Stepping.TmProviders.LocalTm/Stepping/TmProviders/LocalTm/TransactionManagers/LocalTmManager.cs

```diff
@@ -0,0 +1,253 @@
+using Stepping.TmProviders.LocalTm.DistributedLocks;
+using Stepping.TmProviders.LocalTm.Models;
+using Stepping.TmProviders.LocalTm.Steps;
+using Stepping.TmProviders.LocalTm.Store;
+
+namespace Stepping.TmProviders.LocalTm.TransactionManagers;
+
+public class LocalTmManager : ILocalTmManager
+{
+    protected ISteppingDistributedLock DistributedLock { get; }
+
+    protected ILocalTmStore LocalTmStore { get; }
+
+    protected ILocalTmStepExecutor LocalTmStepExecutor { get; }
+
+    protected ISteppingDbContextProviderResolver DbContextProviderResolver { get; }
+
+    protected IBarrierInfoModelFactory BarrierInfoModelFactory { get; }
+
+    protected IDbBarrierInserterResolver DbBarrierInserterResolver { get; }
+
+    protected ILogger<LocalTmManager> Logger { get; }
+
+    public LocalTmManager(
+        ISteppingDistributedLock distributedLock,
+        ILocalTmStore localTmStore,
+        ILocalTmStepExecutor localTmStepExecutor,
+        ISteppingDbContextProviderResolver dbContextProviderResolver,
+        IBarrierInfoModelFactory barrierInfoModelFactory,
+        IDbBarrierInserterResolver dbBarrierInserterResolver,
+        ILogger<LocalTmManager> logger)
+    {
+        DistributedLock = distributedLock;
+        LocalTmStore = localTmStore;
+        LocalTmStepExecutor = localTmStepExecutor;
+        DbContextProviderResolver = dbContextProviderResolver;
+        BarrierInfoModelFactory = barrierInfoModelFactory;
+        DbBarrierInserterResolver = dbBarrierInserterResolver;
+        Logger = logger;
+    }
+
+    public virtual async Task PrepareAsync(string gid, LocalTmStepModel steps, SteppingDbContextLookupInfoModel steppingDbContextLookupInfo,
+        CancellationToken cancellationToken = default)
+    {
+        var tmTransactionModel = new TmTransactionModel(gid, steps, steppingDbContextLookupInfo);
+
+        await CreateAsync(tmTransactionModel, cancellationToken);
+
+        Logger.LogInformation("Local transaction '{gid}' prepared.", gid);
+    }
+
+    public virtual async Task SubmitAsync(string gid, CancellationToken cancellationToken = default)
+    {
+        await using var handle = await DistributedLock.TryAcquireAsync($"LocalTm-{gid}", cancellationToken: cancellationToken);
+
+        if (handle == null)
+        {
+            Logger.LogWarning("Local transaction '{gid}' submit try acquire lock failed.", gid);
+            return;
+        }
+
+        var tmTransactionModel = await GetAsync(gid, cancellationToken);
+
+        if (await IsExpectedStatusAsync(tmTransactionModel, LocalTmConst.StatusPrepare))
+        {
+            return;
+        }
+
+        await UpdateSubmitAsync(tmTransactionModel, cancellationToken);
+
+        Logger.LogInformation("Local transaction '{gid}' committed.", gid);
+    }
+
+    public virtual async Task ProcessPendingAsync(CancellationToken cancellationToken = default)
+    {
+        var pendingTmTransactionModels = await LocalTmStore.GetPendingListAsync(cancellationToken);
```

Comment:
How to truly support multi-instance?

What if we use background (delayed) jobs to process pending?

Reply:
Multi-instance are complex and currently rely only on polling.
Do you have any suggestions?

Reply:
Maybe we can keep the current hosted service processor implementation, and users can use some more professional background workers like Hangfire to replace it.